File: server/src/ai/core/code-parser.ts
import { CodegenType } from '../../common/enums/codegen-type';

export interface ParsedCode {
  files: { filename: string; content: string }[];
}

export class CodeParser {
  private static readonly HTML_REGEX = /```html\s*\n([\s\S]*?)```/i;
  private static readonly CSS_REGEX = /```css\s*\n([\s\S]*?)```/i;
  private static readonly JS_REGEX = /```(?:js|javascript)\s*\n([\s\S]*?)```/i;

  static parse(content: string, codegenType: CodegenType): ParsedCode {
    switch (codegenType) {
      case CodegenType.VANILLA_HTML:
        return CodeParser.parseVanillaHtml(content);
      case CodegenType.MULTI_FILES:
        return CodeParser.parseMultiFiles(content);
      default:
        return { files: [] };
    }
  }

  private static extractCodeByPattern(
    content: string,
    pattern: RegExp,
  ): string {
    const match = content.match(pattern);
    return match ? match[1].trim() : '';
  }

  private static parseVanillaHtml(content: string): ParsedCode {
    const htmlCode = CodeParser.extractCodeByPattern(
      content,
      CodeParser.HTML_REGEX,
    );
    return {
      files: [{ filename: 'index.html', content: htmlCode || content }],
    };
  }

  private static parseMultiFiles(content: string): ParsedCode {
    const files: { filename: string; content: string }[] = [];
    const htmlCode = CodeParser.extractCodeByPattern(
      content,
      CodeParser.HTML_REGEX,
    );
    files.push({ filename: 'index.html', content: htmlCode || content });
    const cssCode = CodeParser.extractCodeByPattern(
      content,
      CodeParser.CSS_REGEX,
    );
    files.push({ filename: 'index.css', content: cssCode || '' });
    const jsCode = CodeParser.extractCodeByPattern(
      content,
      CodeParser.JS_REGEX,
    );
    files.push({ filename: 'index.js', content: jsCode || '' });
    return { files };
  }
}

File: src/main/java/com/github/tianchenghang/core/CodeParser.java
package com.github.tianchenghang.core;

import com.github.tianchenghang.ai.model.MultiFilesResult;
import com.github.tianchenghang.ai.model.VanillaHtmlResult;
import java.util.regex.Pattern;

public class CodeParser {
  private static final Pattern HTML_CODE_PATTERN =
      Pattern.compile("```html\\s*\\n([\\s\\S]*?)```", Pattern.CASE_INSENSITIVE);
  private static final Pattern CSS_CODE_PATTERN =
      Pattern.compile("```css\\s*\\n([\\s\\S]*?)```", Pattern.CASE_INSENSITIVE);
  private static final Pattern JS_CODE_PATTERN =
      Pattern.compile("```(?:js|javascript)\\s*\\n([\\s\\S]*?)```", Pattern.CASE_INSENSITIVE);

  public static VanillaHtmlResult parseHtmlCode(String codeContent) {
    var result = new VanillaHtmlResult();
    var htmlCode = extractHtmlCode(codeContent);
    if (htmlCode != null && !htmlCode.trim().isEmpty()) {
      result.setHtmlCode(htmlCode);
    } else {
      result.setHtmlCode(codeContent.trim());
    }
    return result;
  }

  public static MultiFilesResult parseMultiFiles(String codeContent) {
    var result = new MultiFilesResult();
    var htmlCode = extractCodeByPattern(codeContent, HTML_CODE_PATTERN);
    var cssCode = extractCodeByPattern(codeContent, CSS_CODE_PATTERN);
    var jsCode = extractCodeByPattern(codeContent, JS_CODE_PATTERN);
    if (htmlCode != null && !htmlCode.trim().isEmpty()) {
      result.setHtmlCode(htmlCode);
    }
    if (cssCode != null && !cssCode.trim().isEmpty()) {
      result.setCssCode(cssCode);
    }
    if (jsCode != null && !jsCode.trim().isEmpty()) {
      result.setJsCode(jsCode);
    }
    return result;
  }

  private static String extractHtmlCode(String content) {
    var matcher = HTML_CODE_PATTERN.matcher(content);
    if (matcher.find()) {
      return matcher.group(1);
    }
    return null;
  }

  private static String extractCodeByPattern(String content, Pattern pattern) {
    var matcher = pattern.matcher(content);
    if (matcher.find()) {
      return matcher.group(1);
    }
    return null;
  }
}
